fusedrug_examples/interaction/drug_target/affinity_prediction/t5_mol/tokenizer/multi_tokenizer.py

```diff
@@ -0,0 +1,1304 @@
+            )
+
+        return sample_dict
+
+
+class ModularStringTokenizerOp(OpBase):
+    """
+    Tokenizes a raw input string containing multiple types of molecule representations (e.g. AA sequences, SMILES, SELFIES, etc.),
+    by inferring which parts of the string correspond to which tokenizer, and applying the corresponding tokenizers.
+    applies huggingface(https://github.com/huggingface/tokenizers)-based tokenizers
+    """
+
+    def __init__(
+        self,
+        tokenizer_dict: Dict,  # a dictionary of tokenizer instances whose keys are the input types to be transferred to _call_
+        # Every value is a dict containing the following:
+        # "tokenizer_inst": tokenizer instance
+        # "max_size": max number of output tokens
+        max_size: Optional[int] = None,  # determines both padding length and max size.
+        pad_id: Optional[int] = None,
+        pad_type_id: Optional[int] = None,
+        verbose: Optional[int] = 0,
+        **kwargs: Any,
+    ) -> None:
+        """
+
+        Args:
+            tokenizer_dict (Dict): a dictionary of tokenizers, keyed by input type (i.e. to which input to apply each tokenizer).
+            The tokenizers must map to a single ID space, i.e.
+            1. different tokens from each tokenizer must map to different IDs
+            2. All tokenizers have the same special tokens, with the same IDs
+            3. Same tokens from different tokenizers may map to different IDs
+            Every value is a dict containing the following:
+                "tokenizer_inst": tokenizer instance
+                "max_size": max number of output tokens
+                "start_delimiter": starting delimiter of an input sequence that is to be tokenized by the given tokenizer
+                "end_delimiter": ending delimiter of an input sequence that is to be tokenized by the given tokenizer
+            pad_type_id (_type_, optional): _description_. Defaults to None.
+            verbose (int, optional): _description_. Defaults to 0.
+        """
+        super().__init__(**kwargs)
+
+        raise Exception("Not implemented")
+
+    def get_token_id(self, token_str: str) -> int:
+        """returns the id the token maps to
+
+        Args:
+            token_str (str): _description_
+
+        Raises:
+            Exception: if could not find the token in any of the internal tokenizers
+
+        Returns:
+            int: ID of the input token if it exists
+        """
+        raise Exception("Not implemented")
+
+    def __call__(
+        self,
+        sample_dict: NDict,
+        key_in: str,
+        key_out_tokenized_object: str = None,
+        key_out_tokens_ids: str = None,
+        key_out_attention_mask: str = None,
+        convert_attention_mask_to_bool: bool = True,
+        input_type: Optional[str] = None,
+    ) -> NDict:
+        raise Exception("Not implemented")
+
+
+class ModularMultiDecoderOp(OpBase):
```

Comment:
Not sure if we need Op for that since we probably will use it on model output (after the data pipeline).

Reply:
Removed it.